raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -109,50 +114,86 @@ private OptionalLong quorumMajorityFetchTimestamp() {
+        return new ArrayList<>(this.voterReplicaStates.values())
+            .stream().sorted().map(voterState -> new ReplicaState()
+                .setReplicaId(voterState.nodeId)
+                .setLastCaughtUpTimeMs(voterState.lastFetchTimestamp.orElse(-1))
+                .setLogEndOffset(voterState.endOffset.orElse(-1)))
+                .collect(Collectors.toList());
     }
 
     /**
      * @return true if the high watermark is updated too
      */
     public boolean updateEndOffset(int remoteNodeId, long endOffset) {
-        ReplicaState state = ensureValidVoter(remoteNodeId);
+        NodeState state = getReplicaState(remoteNodeId);
+
         state.endOffset.ifPresent(currentEndOffset -> {
             if (currentEndOffset > endOffset)
                 throw new IllegalArgumentException("Non-monotonic update to end offset for nodeId " + remoteNodeId);
         });
-        state.hasEndorsedLeader = true;
         state.endOffset = OptionalLong.of(endOffset);
-        return updateHighWatermark();
+        if (isVoter(remoteNodeId)) {
+            addEndorsementFrom(remoteNodeId);
+            return updateHighWatermark();
+        }
+        return false;
     }
 
     public void addEndorsementFrom(int remoteNodeId) {
-        ReplicaState replicaState = ensureValidVoter(remoteNodeId);
-        replicaState.hasEndorsedLeader = true;
+        VoterState voterState = ensureValidVoter(remoteNodeId);
+        voterState.hasEndorsedLeader = true;
     }
 
-    private ReplicaState ensureValidVoter(int remoteNodeId) {
-        ReplicaState state = voterReplicaStates.get(remoteNodeId);
+    private VoterState ensureValidVoter(int remoteNodeId) {
+        VoterState state = voterReplicaStates.get(remoteNodeId);
         if (state == null)
             throw new IllegalArgumentException("Unexpected endorsement from non-voter " + remoteNodeId);
         return state;
     }
 
+    NodeState getReplicaState(int remoteNodeId) {
+        NodeState state = voterReplicaStates.get(remoteNodeId);
+        if (state == null) {
+            observerReplicaStates.putIfAbsent(remoteNodeId, new NodeState(remoteNodeId));
+            return observerReplicaStates.get(remoteNodeId);
+        }
+        return state;
+    }
+
+    List<ReplicaState> getVoterStates() {
+        return voterReplicaStates.entrySet().stream().filter(entry -> entry.getKey() != localId)
+            .map(entry -> new ReplicaState()
+                              .setReplicaId(entry.getKey())
+                              .setLogEndOffset(entry.getValue().endOffset.orElse(-1)))
+            .collect(Collectors.toList());
+    }
+
+    List<ReplicaState> getObserverStates() {
+        return observerReplicaStates.entrySet().stream()
+                   .map(entry -> new ReplicaState()
+                                     .setReplicaId(entry.getKey())
+                                     .setLogEndOffset(entry.getValue().endOffset.orElse(-1)))
```

Comment:
We are not setting last caught up time. I guess we need to add the bookkeeping for it. You can look at `kafka.cluster.Replica` to see how we track it for regular partitions. We could also decide that it is not very useful and leave it out of the API.